light/client.go

```diff
@@ -729,45 +723,74 @@ func (c *Client) verifySkippingAgainstPrimary(
 	now time.Time) error {
 
 	trace, err := c.verifySkipping(ctx, c.primary, trustedBlock, newLightBlock, now)
+	if err == nil {
+		// Success! Now compare the header with the witnesses to ensure it's not a fork.
+		// More witnesses we have, more chance to notice one.
+		//
+		// CORRECTNESS ASSUMPTION: there's at least 1 correct full node
+		// (primary or one of the witnesses).
+		if cmpErr := c.detectDivergence(ctx, trace, now); cmpErr != nil {
+			return cmpErr
+		}
+	}
+
+	// all errors from verify skipping should be `ErrVerificationFailed`
+	e, ok := err.(ErrVerificationFailed)
+	if !ok {
+		return err
+	}
 
-	switch errors.Unwrap(err).(type) {
+	replace := true
+	switch e.Reason.(type) {
+	// Verification returned an invalid header
 	case ErrInvalidHeader:
-		// If the target header is invalid, return immediately.
-		invalidHeaderHeight := err.(ErrVerificationFailed).To
-		if invalidHeaderHeight == newLightBlock.Height {
+		// If it was the target header, return immediately.
+		if e.To == newLightBlock.Height {
 			c.logger.Debug("target header is invalid", "err", err)
 			return err
 		}
 
-		// If some intermediate header is invalid, remove the primary and try again.
-		c.logger.Error("primary sent invalid header -> replacing", "err", err, "primary", c.primary)
-		replacementBlock, removeErr := c.findNewPrimary(ctx, newLightBlock.Height, true)
-		if removeErr != nil {
-			c.logger.Error("failed to replace primary. Returning original error", "err", removeErr)
-			return err
-		}
+		// If some intermediate header is invalid, remove the primary and try
+		// again.
 
-		if !bytes.Equal(replacementBlock.Hash(), newLightBlock.Hash()) {
-			c.logger.Debug("replaced primary but new primary has a different block to the initial one. Returning original error")
-			return err
+	// An intermediate header expired. We can no longer validate it as there is
+	// no longer the ability to punish invalid blocks as evidence of misbehavior
+	case ErrOldHeaderExpired:
+		return err
+
+	// This happens if there was a problem in finding the next block or a
+	// context was canceled.
+	default:
+		if errors.Is(e.Reason, context.Canceled) || errors.Is(e.Reason, context.DeadlineExceeded) {
+			return e.Reason
 		}
 
-		// attempt to verify the header again
-		return c.verifySkippingAgainstPrimary(ctx, trustedBlock, replacementBlock, now)
-	case nil:
-		// Compare header with the witnesses to ensure it's not a fork.
-		// More witnesses we have, more chance to notice one.
-		//
-		// CORRECTNESS ASSUMPTION: there's at least 1 correct full node
-		// (primary or one of the witnesses).
-		if cmpErr := c.detectDivergence(ctx, trace, now); cmpErr != nil {
-			return cmpErr
+		if !provider.ShouldBeRemoved(e.Reason) {
+			replace = false
 		}
-	default:
-		return err
 	}
 
-	return nil
+	// if we've reached here we're attempting to retry verification with a
+	// different provider
+	if replace {
```

Comment:
This seems awkward and it sort of seems like we're logging at the wrong level, because I can't see where we're doing the replacement operation (I trust it's in findNewPrimary), it seems like we could either: log there, or just log it all at info. The error case, because we handle it doesn't seem worth logging at error level.

Reply:
Yeah that totally makes sense. I think I should just have it within the `findNewPrimary` function

Reply:
Is the fact that we handle an error (by replacing the faulty primary), sufficient enough to warrant keeping the log level at `Info`?